This is the dataset of Nepal Census 2021..


In [251]:
import numpy as np
import pandas as pd

In [252]:
data=pd.read_csv("census.csv")

In [253]:
data

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female
0,1,Taplejung,Aathrai Tribeni Rural Municipality,2869,2735,12288,6005,6283
1,2,Taplejung,Maiwakhola Rural Municipality,2275,2178,10365,5264,5101
2,3,Taplejung,Meringden Rural Municipality,2683,2528,12040,6181,5859
3,4,Taplejung,Mikwakhola Rural Municipality,1862,1792,7991,4000,3991
4,5,Taplejung,Phaktanglung Rural Municipality,2864,2700,11925,6239,5686
...,...,...,...,...,...,...,...,...
748,749,Darchula,Mahakali Municipality,6080,4596,24572,12072,12500
749,750,Darchula,Malikaarjun Rural Municipality,3241,2993,15754,7631,8123
750,751,Darchula,Marma Rural Municipality,3086,2704,15586,7551,8035
751,752,Darchula,Naugad Rural Municipality,3156,2920,16434,7988,8446


Data Understanding (5 questions)
1. What are the shape (rows & columns), data types, and basic info of the dataset?

In [254]:
data.shape

(753, 8)

our data set have 753 rows and 8 columns

In [255]:
data.columns

Index(['_id', 'District', 'Local Level Name', 'Total family number',
       'Total household number', 'Total population', 'Total Male',
       'Total Female'],
      dtype='object')

these are our columns:

id = id number 

District = Name of the district

Local Level Name = name of the municipality under the district

Total family number = Total number of families

Total household number = total number of houses 

Total population = total number of people

Total Male = total number of male population

Total Female = Total number of female population

In [256]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 753 entries, 0 to 752
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   _id                     753 non-null    int64 
 1   District                753 non-null    object
 2   Local Level Name        753 non-null    object
 3   Total family number     753 non-null    int64 
 4   Total household number  753 non-null    int64 
 5   Total population        753 non-null    int64 
 6   Total Male              753 non-null    int64 
 7   Total Female            753 non-null    int64 
dtypes: int64(6), object(2)
memory usage: 47.2+ KB


2. How many unique Districts and Local Levels are there?

In [257]:
data['District'].unique()

array(['Taplejung', 'Panchthar', 'Ilam', 'Jhapa', 'Morang', 'Sunsari',
       'Dhankuta', 'Terhathum', 'Sankhuwasabha', 'Bhojpur', 'Solukhumbu',
       'Okhaldhunga', 'Khotang', 'Udayapur', 'Saptari', 'Siraha',
       'Dhanusa', 'Mahottari', 'Sarlahi', 'Rautahat', 'Bara', 'Parsa',
       'Sindhuli', 'Ramechhap', 'Dolakha', 'Sindhupalchok',
       'Kavrepalanchok', 'Lalitpur', 'Bhaktapur', 'Kathmandu', 'Nuwakot',
       'Rasuwa', 'Dhading', 'Makwanpur', 'Chitawan', 'Gorkha', 'Lamjung',
       'Tanahu', 'Syangja', 'Kaski', 'Manang', 'Mustang', 'Myagdi',
       'Parbat', 'Baglung', 'Nawalparasi East', 'Gulmi', 'Palpa',
       'Nawalparasi West', 'Rupandehi', 'Kapilbastu', 'Arghakhanchi',
       'Pyuthan', 'Rolpa', 'Rukum East', 'Dang', 'Banke', 'Bardiya',
       'Rukum West', 'Salyan', 'Surkhet', 'Dailekh', 'Jajarkot', 'Dolpa',
       'Jumla', 'Kalikot', 'Mugu', 'Humla', 'Bajura', 'Bajhang', 'Achham',
       'Doti', 'Kailali', 'Kanchanpur', 'Dadeldhura', 'Baitadi',
       'Darchula'], dty

In [258]:
len(data['District'].unique())


77

77 unique districts

In [259]:
data["Local Level Name"].unique()

array(['Aathrai Tribeni Rural Municipality',
       'Maiwakhola Rural Municipality', 'Meringden Rural Municipality',
       'Mikwakhola Rural Municipality', 'Phaktanglung Rural Municipality',
       'Phungling Municipality', 'Sidingba Rural Municipality',
       'Sirijangha Rural Municipality',
       'Pathivara Yangwarak Rural Municipality',
       'Falelung Rural Municipality', 'Falgunanda Rural Municipality',
       'Hilihang Rural Municipality', 'Kummayak Rural Municipality',
       'Miklajung Rural Municipality', 'Phidim Municipality',
       'Tumbewa Rural Municipality', 'Yangwarak Rural Municipality',
       'Chulachuli Rural Municipality', 'Deumai Municipality',
       'Fakphokthum Rural Municipality', 'Illam Municipality',
       'Maijogmai Rural Municipality', 'Mai Municipality',
       'Mangsebung Rural Municipality', 'Rong Rural Municipality',
       'Sandakpur Rural Municipality', 'Suryodaya Municipality',
       'Arjundhara Municipality', 'Barhadashi Rural Municipality',


In [260]:
len(data['Local Level Name'].unique())

737

In [261]:
data['Local Level Name'].nunique()

737

737 unique Local level names.

3. Are there any missing values or duplicate rows in the dataset?

In [262]:
data.isnull().sum()

_id                       0
District                  0
Local Level Name          0
Total family number       0
Total household number    0
Total population          0
Total Male                0
Total Female              0
dtype: int64

No, there is no any missing values

In [263]:
data.duplicated().sum()

np.int64(0)

No, their are no duplicate rows.

4. What is the summary statistics (mean, median, min, max, std) for numerical columns like Total Population, Total Male, Total Female, etc.?

In [264]:
data.describe()

,_id,Total family number,Total household number,Total population,Total Male,Total Female
count,753.000000,753.000000,753.000000,753.000000,753.000000,753.000000
mean,377.000000,8976.723772,7493.590969,38612.204515,18842.814077,19769.390438
std,217.516666,12986.552238,8401.892705,49904.377212,24891.928559,25046.362147
min,1.000000,125.000000,125.000000,398.000000,170.000000,228.000000
25%,189.000000,4036.000000,3653.000000,16768.000000,8115.000000,8553.000000
50%,377.000000,5896.000000,5195.000000,25508.000000,12248.000000,13010.000000
75%,565.000000,9510.000000,8151.000000,44929.000000,22100.000000,23068.000000
max,753.000000,231714.000000,105649.000000,845767.000000,431501.000000,414266.000000


5. What is the range and distribution of Total Family Number and Total Household Number?

In [265]:
# to find range 
min_family_num = data['Total family number'].min()
max_family_num = data['Total family number'].max()
family_range = max_family_num - min_family_num
print(f"Range of Total family number = {family_range}")

min_household = data['Total household number'].min()
max_household = data['Total household number'].max()
household_range = max_household - min_household
print(f"Range of Total household number = {household_range}")

Range of Total family number = 231589
Range of Total household number = 105524


In [266]:
# to find distribution
data['Total family number'].describe()

count       753.000000
mean       8976.723772
std       12986.552238
min         125.000000
25%        4036.000000
50%        5896.000000
75%        9510.000000
max      231714.000000
Name: Total family number, dtype: float64

In [267]:
data["Total household number"].describe()

count       753.000000
mean       7493.590969
std        8401.892705
min         125.000000
25%        3653.000000
50%        5195.000000
75%        8151.000000
max      105649.000000
Name: Total household number, dtype: float64

data Cleaning & Manipulation (5 questions)

6. Verify data consistency: Does Total Male + Total Female equal Total Population for all rows? If not, how many mismatches?

In [268]:
data['mismatch']=(data['Total Male'] + data['Total Female']) == data['Total population']
data

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female,mismatch
0,1,Taplejung,Aathrai Tribeni Rural Municipality,2869,2735,12288,6005,6283,True
1,2,Taplejung,Maiwakhola Rural Municipality,2275,2178,10365,5264,5101,True
2,3,Taplejung,Meringden Rural Municipality,2683,2528,12040,6181,5859,True
3,4,Taplejung,Mikwakhola Rural Municipality,1862,1792,7991,4000,3991,True
4,5,Taplejung,Phaktanglung Rural Municipality,2864,2700,11925,6239,5686,True
...,...,...,...,...,...,...,...,...,...
748,749,Darchula,Mahakali Municipality,6080,4596,24572,12072,12500,True
749,750,Darchula,Malikaarjun Rural Municipality,3241,2993,15754,7631,8123,True
750,751,Darchula,Marma Rural Municipality,3086,2704,15586,7551,8035,True
751,752,Darchula,Naugad Rural Municipality,3156,2920,16434,7988,8446,True


In [269]:
data[data['mismatch']=='False']

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female,mismatch


No, There is no mismatch in the total population. Total Male + Total Female = Total population

7. Create new columns: Population Density proxy (if possible), Male Percentage, Female Percentage, and Sex Ratio (Male:Female).

In [270]:
data['Male percentage'] = ((data['Total Male'] / data['Total population']) * 100).round(2)
data['Female percentage'] = ((data['Total Female'] / data['Total population']) * 100).round(2)
data['Sex Ration'] = (data['Total Male'] / data['Total Female']).round(2)

# Population Density (Average population per household)
data['population density'] = data['Total population'] / data['Total household number']


In [271]:
data.sample(10)

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female,mismatch,Male percentage,Female percentage,Sex Ration,population density
311,312,Kavrepalanchok,Banepa Municipality,16973,12545,67629,33180,34449,True,49.06,50.94,0.96,5.390913
516,517,Rupandehi,Sainamaina Municipality,19657,17572,78477,36547,41930,True,46.57,53.43,0.87,4.466025
232,233,Rautahat,Gaur Municipality,7743,6062,40368,20655,19713,True,51.17,48.83,1.05,6.659188
377,378,Makwanpur,Bhimphedi Rural Municipality,5128,4939,21626,10668,10958,True,49.33,50.67,0.97,4.378619
641,642,Jumla,Patrasi Rural Municipality,3374,3191,16881,8349,8532,True,49.46,50.54,0.98,5.290191
229,230,Rautahat,Durga Bhagwati Rural Municipality,4136,3369,23226,11809,11417,True,50.84,49.16,1.03,6.894034
366,367,Dhading,Gangajamuna Rural Municipality,5305,5183,19460,9494,9966,True,48.79,51.21,0.95,3.754582
714,715,Kailali,Kailari Rural Municipality,10793,10458,50457,24485,25972,True,48.53,51.47,0.94,4.824727
234,235,Rautahat,Ishanath Municipality,8084,6289,53402,26821,26581,True,50.22,49.78,1.01,8.491334
618,619,Dailekh,Mahabu Rural Municipality,4021,3652,18037,8316,9721,True,46.11,53.89,0.86,4.938938


8. Group the data by District and calculate aggregate totals (Total Population, Households, Families per district).

In [272]:
district_summary = data.groupby('District').agg({
    'Total population' : 'sum',
    'Total household number': 'sum',
    'Total family number':'sum'
}).reset_index()
district_summary

,District,Total population,Total household number,Total family number
0,Achham,228907,39368,49962
1,Arghakhanchi,176534,45012,49073
2,Baglung,248964,58445,65334
3,Baitadi,244400,42288,50472
4,Bajhang,189023,32916,38368
...,...,...,...,...
72,Syangja,254277,62856,70710
73,Tanahu,326962,81642,93168
74,Taplejung,120359,26139,28530
75,Terhathum,88885,20352,22022


In [273]:
district_summary[district_summary['District']=='Bara']

,District,Total population,Total household number,Total family number
7,Bara,742667,112301,134362


9. Identify and handle any outliers in Total Population or Household counts.

In [274]:
# outliers for Total population
# Calculate Q1 (25th percentile) and Q3 (75th percentile)
population_q1 = data['Total population'].quantile(0.25)
population_q3 = data['Total population'].quantile(0.75)

# calculate the IQR (Inter Quantile Range)
population_IQR = q3-q1

# Define outlier boundariesa
population_lower_bound = population_q1 - 1.5 * population_IQR
population_upper_bound = population_q3 + 1.5 * population_IQR

# filtering the rows containing outliers
population_outliers = data[(data['Total population'] < population_lower_bound) | (data['Total population'] > population_upper_bound)]
print(len(population_outliers))

population_median = data['Total population'].median()
# dataFrame.loc[condition, 'Column_Name'] = 'Value'

data.loc[population_outliers.index,'Total population']=population_median

# Check if any value still exceeds your ORIGINAL upper bound
remaining_population_outliers = data[((data['Total population'] > population_upper_bound) | (data['Total population']< population_lower_bound))]
print(f"population remaining outliers : {len(remaining_extreme_outliers)}") 
# This will output: 0

60
population remaining outliers : 0


In [275]:
# outliers for Total household number
# Calculate Q1 (25th percentile) and Q3 (75th percentile)
household_q1 = data['Total household number'].quantile(0.25)
household_q3 = data['Total household number'].quantile(0.75)

# calculate the IQR (Inter Quantile Range)
household_IQR = household_q3 - household_q1

# Define outlier boundaries
household_lower_bound = household_q1 - 1.5 * household_IQR
household_upper_bound = household_q3 +1.5 * household_IQR

# filtering the rows containing outliers
household_outliers = data[(data['Total household number']< household_lower_bound) | (data['Total household number']>household_upper_bound)]
print(len(household_outliers))
household_median = data['Total household number'].median()
# dataFrame.loc[condition, 'Column_Name'] = 'Value'
data.loc[household_outliers.index,'Total household number']=household_median

# Check if any value still exceeds your ORIGINAL upper bound
remaining_household_outliers = data[((data['Total population'] > household_upper_bound) | (data['Total population']< household_lower_bound))]
print(f"household remaining outliers: {len(remaining_extreme_outliers)}") 

65
household remaining outliers: 0


In [277]:
data

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female,mismatch,Male percentage,Female percentage,Sex Ration,population density
0,1,Taplejung,Aathrai Tribeni Rural Municipality,2869,2735,12288,6005,6283,True,48.87,51.13,0.96,4.492870
1,2,Taplejung,Maiwakhola Rural Municipality,2275,2178,10365,5264,5101,True,50.79,49.21,1.03,4.758953
2,3,Taplejung,Meringden Rural Municipality,2683,2528,12040,6181,5859,True,51.34,48.66,1.05,4.762658
3,4,Taplejung,Mikwakhola Rural Municipality,1862,1792,7991,4000,3991,True,50.06,49.94,1.00,4.459263
4,5,Taplejung,Phaktanglung Rural Municipality,2864,2700,11925,6239,5686,True,52.32,47.68,1.10,4.416667
...,...,...,...,...,...,...,...,...,...,...,...,...,...
748,749,Darchula,Mahakali Municipality,6080,4596,24572,12072,12500,True,49.13,50.87,0.97,5.346388
749,750,Darchula,Malikaarjun Rural Municipality,3241,2993,15754,7631,8123,True,48.44,51.56,0.94,5.263615
750,751,Darchula,Marma Rural Municipality,3086,2704,15586,7551,8035,True,48.45,51.55,0.94,5.764053
751,752,Darchula,Naugad Rural Municipality,3156,2920,16434,7988,8446,True,48.61,51.39,0.95,5.628082


10. Create a new feature: Average Family Size (Total Population / Total Family Number) and Average Household Size.

In [284]:
data['Average family size'] = (data['Total population'] / data['Total family number']).round(2)
data['Average household size'] = (data['Total population'] / data['Total household number']).round(2)
data

,_id,District,Local Level Name,Total family number,Total household number,Total population,Total Male,Total Female,mismatch,Male percentage,Female percentage,Sex Ration,population density,Average family size,Average household size
0,1,Taplejung,Aathrai Tribeni Rural Municipality,2869,2735,12288,6005,6283,True,48.87,51.13,0.96,4.492870,4.28,4.49
1,2,Taplejung,Maiwakhola Rural Municipality,2275,2178,10365,5264,5101,True,50.79,49.21,1.03,4.758953,4.56,4.76
2,3,Taplejung,Meringden Rural Municipality,2683,2528,12040,6181,5859,True,51.34,48.66,1.05,4.762658,4.49,4.76
3,4,Taplejung,Mikwakhola Rural Municipality,1862,1792,7991,4000,3991,True,50.06,49.94,1.00,4.459263,4.29,4.46
4,5,Taplejung,Phaktanglung Rural Municipality,2864,2700,11925,6239,5686,True,52.32,47.68,1.10,4.416667,4.16,4.42
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
748,749,Darchula,Mahakali Municipality,6080,4596,24572,12072,12500,True,49.13,50.87,0.97,5.346388,4.04,5.35
749,750,Darchula,Malikaarjun Rural Municipality,3241,2993,15754,7631,8123,True,48.44,51.56,0.94,5.263615,4.86,5.26
750,751,Darchula,Marma Rural Municipality,3086,2704,15586,7551,8035,True,48.45,51.55,0.94,5.764053,5.05,5.76
751,752,Darchula,Naugad Rural Municipality,3156,2920,16434,7988,8446,True,48.61,51.39,0.95,5.628082,5.21,5.63
